### PR TITLE
Pr/hipsycl

### PR DESCRIPTION
**Pull request by bd4**
This only works so far with the omp backend, but it gives an idea of some of the issues we need to work through.

### CODE REVIEW

File: include/sycl_ext_complex.hpp [hunk 1]
```diff
@@ -246,9 +246,27 @@ template<class T> complex<T> tanh (const complex<T>&);
 
 #define _SYCL_EXT_CPLX_BEGIN_NAMESPACE_STD namespace sycl::ext::cplx {
 #define _SYCL_EXT_CPLX_END_NAMESPACE_STD }
+
+#ifdef __HIPSYCL__
+
+#define _SYCL_EXT_CPLX_INLINE_VISIBILITY inline
+#define SYCL_EXTERNAL
```
[the remark this answers is not in the document]
Comment (illuhad) [reply include/sycl_ext_complex.hpp:253]: Looking through the code, I don't understand why the header even uses `SYCL_EXTERNAL`. 
What it does (if supported) is that it allows you to have definitions of device code in a different compilation unit. I don't think the code here needs that.

[the remark this answers is not in the document]
Comment (TApplencourt) [reply include/sycl_ext_complex.hpp:253]: > SYCL_EXTERNAL is an optional macro which enables external linkage of SYCL functions and member functions to be included in a SYCL kernel.

It's optional you are 100 right. 

> What it does (if supported) is that it allows you to have definitions of device code in a different compilation unit. I don't think the code here needs that.

Just to be sure I follow you (sorry, I'm slow). We need `SYCL_EXTERNAL` when the definition and the use and in two different translation unit. But we are in a header file, so by definition, we will be on the same translation unit, so we don't need to decore our function with `SYCL_EXTERNAL`. 
Is this correct?

[the remark this answers is not in the document]
Comment (illuhad) [reply include/sycl_ext_complex.hpp:253]: Precisely. A header-only SYCL library would never need `SYCL_EXTERNAL`. You would only need it if you had the function definitions in a separate cpp file :)

[the remark this answers is not in the document]
Comment (TApplencourt) [reply include/sycl_ext_complex.hpp:253]: I'm convinced! Will remove SYCL_EXTERNAL :)

**Comment by illuhad** [comment]
@TApplencourt It's true that the spec reserves this namespace for extensions. However, the spec also assumes that extensions are shipped as part of the code of one implementation, and therefore ineherently implementation-specific. Because of this, extensions are assumed to be able to make assumptions about the internals of implementations.

I guess the core question is whether you just want have a DPC++-specific extension, or whether you want to have a generic SYCL library. In the latter case, you are operating at the same level as a regular user library, and then making assumptions about implementation internals is not a good idea :)

To give some more motivation about the issue at hand: Imagine somebody included the legacy `#include <CL/sycl.hpp>` header. In that case, the extension would have to also pop up under `cl::sycl::ext::cplx`. It is not guaranteed how this namespace interacts with `::sycl`, so just putting everything in `::sycl:.ext`might not work.

If you want to support multiple SYCL implementations, I see two options. You can either
* Have implementation-specific code paths that assume the internals of individual implementations. E.g: 
```c++
#ifdef __HIPSYCL__
 #define _SYCL_EXT_CPLX_BEGIN_NAMESPACE_STD namespace hipsycl::sycl::ext::cplx {
#else
  #define _SYCL_EXT_CPLX_BEGIN_NAMESPACE_STD namespace sycl::ext::cplx {
#endif
```

* Not put things into `sycl` , and work with your own namespace (e.g. `syclcomplex`) like a regular user library would.

**Comment by TApplencourt** [comment]
>I guess the core question is whether you just want to have a DPC++-specific extension, or whether you want to have a generic SYCL library.

So we don't want to have a DPC++specific extension for sure! :)  At some point, we want to merge `sycl::complex` into the spec, and then this lib should be provided by Khronos so implementers can use them if they want to support `sycl` complex. 

I guess we can do something like (credit to @Kerilk):

```c++
#ifndef _SYCL_CPLX_NAMESPACE
#ifdef __HIPSYCL__
 #define _SYCL_CPLX_NAMESPACE hipsycl::sycl::ext
#else
#define _SYCL_CPLX_NAMESPACE sycl::ext
#endif
#endif

#define _SYCL_EXT_CPLX_BEGIN_NAMESPACE_STD SYCL_CPLX_NAMESPACE::cplx {
```

So user / implemented can always overwrite their namespace. But if they don't, we peggy back on the heuristic. And when it's merged in the spec (hopefully) we delete all the macro \o/

**Comment by illuhad** [comment]
That sounds like a good approach!

**Comment by bd4** [comment]
Yes #22 replaces this. hipSYCL still doesn't have all of the helpers in `sycl` namespace, so still need this for now to experiment:
```
#include <sycl/sycl.hpp>                                                                 
                                                                                         
namespace hipsycl::sycl {                                                                
using std::isinf;                                                                        
using std::isfinite;                                                                     
using std::ldexp;                                                                        
using std::signbit;                                                                      
} // namespace hipsycl::sycl
                                                        
#include "sycl_ext_complex.hpp"
```

**Pull request closed without merging**